maven.nix

```diff
@@ -1,23 +1,94 @@
-{ lib, fetchurl, linkFarm }:
-with lib;
-# Create a maven environment from the output of the mvn2nix command
-# the resulting store path can be used as the a .m2 repository for subsequent
-# maven invocations.
-# ex.
-# 	mvn package --offline -Dmaven.repo.local=${dependencies}
-#
-# @param name: The optional name of the derivation
-# @param dependencies: The nix expression generated by mvn2nix
-{ name ? "buildMavenRepository", dependencies }:
-let
-  dependenciesAsDrv = (forEach (attrValues dependencies) (dependency: {
-    drv = fetchurl {
-      url = dependency.url;
-      sha256 = dependency.sha256;
+{ pkgs, makeWrapper, stdenv, lib, fetchurl, linkFarm }:
+with lib; rec {
+  # Create a maven environment from the output of the mvn2nix command
+  # the resulting store path can be used as the a .m2 repository for subsequent
+  # maven invocations.
+  # ex.
+  # 	mvn package --offline -Dmaven.repo.local=${dependencies}
+  #
+  # @param generated: The nix expression generated by mvn2nix
+  buildMavenRepository = { generated }:
+    let
+      dependenciesAsDrv = (forEach (attrValues generated.dependencies)
+        (dependency: {
+          drv = fetchurl {
+            url = dependency.url;
+            sha256 = dependency.sha256;
+          };
+          layout = dependency.layout;
+        }));
+    in linkFarm "buildMavenRepository" (forEach dependenciesAsDrv (dependency: {
+      name = dependency.layout;
+      path = dependency.drv;
+    }));
+
+  # Execute maven and build the project.
+  # A wrapper script will be created in the `bin` folder to execute the JAR.
+  #
+  # Make sure that the JAR is setup to search for classpaths in a repository layout.
+  # ex.
+  #             <plugin>
+  #                  <artifactId>maven-jar-plugin</artifactId>
+  #                  <version>3.2.0</version>
+  #                  <configuration>
+  #                      <archive>
+  #                          <manifest>
+  #                              <addClasspath>true</addClasspath>
+  #                              <classpathPrefix>lib/</classpathPrefix>
+  #                              <classpathLayoutType>repository</classpathLayoutType>
+  #                              <mainClass>com.fzakaria.mvn2nix.Main</mainClass>
+  #                          </manifest>
+  #                          <manifestEntries>
+  #                              <Class-Path>.</Class-Path>
+  #                          </manifestEntries>
+  #                      </archive>
+  #                  </configuration>
+  #              </plugin>
+  #
+  # @param maven: The maven version to use; defaults to the one pinned by mvn2nix
+  # @param jdk: The JDK version to use; defaults to the one pinned by mvn2nix
+  # @param generated: The nix expression generated by mvn2nix
+  # @param repository: A maven repository; see buildMavenRepository above.
+  # @param src: The src files to feed into maven
+  buildMaven = { maven ? pkgs.maven, jdk ? pkgs.jdk, generated
```

Comment:
Might make sense to split this:
`buildMavenJar` & `buildMavenExecutable` or something and have `buildMavenExecutable` create a `linkFarm` of the repository & output of `buildMavenJar`